roadmap/implementers-guide/src/types/messages.md

```diff
@@ -26,3 +29,34 @@ struct UpwardMessage {
 	pub data: Vec<u8>,
 }
 ```
+
+## Horizontal Message
+
+This is a message sent from a parachain to another parachain that travels through the relay chain.
+This message ends up in the recipient's mailbox. A size of a horizontal message is defined by its
+`data` payload.
+
+```rust,ignore
+struct HorizontalMessage {
+	/// The para that will get this message in its downward message queue.
+	pub recipient: ParaId,
+	/// The message payload.
+	pub data: Vec<u8>,
+}
+```
+
+## Downward Message
+
+A message that go down from the relay chain to a parachain. Such a message could be initiated either
+as a result of an operation took place on the relay chain or sent using a horizontal message.
+
+```rust,ignore
+enum DownwardMessage {
```

Comment:
The `Opaque` variant is for governance and stuff to send messages directly to parachains. Also, I think @gavofyork plans to use it to offload relay-chain functionality onto relay chains. So then there will be a special protocol between the relay-chain and the parachain that will use these `Opaque` messages to communicate.

Reply:
Ok, got it. I couldn't help myself to choose the name `ParachainSpecific` since `Opaque` variant is only opaque for third-party paras who don't participate in this custom protocol: both relay-chain and the recipient para know about the contents. But my reasoning is weak, will be happy to revert to `Opaque`.

Reply:
`ParachainSpecific` is more clear to me.